clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -687,7 +687,6 @@ public CompletableFuture<Void> leaveGroup() {
                     "to clear its assignment and send a leave group heartbeat", memberId);
             }
             // Clear the subscription, no matter if the callback execution failed or succeeded.
-            subscriptions.unsubscribe();
```

Comment:
This change means `leaveGroup` does NOT cleanup any subscribed/assigned partitions from `SubscriptionState`. It seems to me that is a bit weird since that is `leaveGroup`'s  responsibility to ensure the cleanup of subscribed/assigned partitions.  Maybe another solution is to add `subscriptions.unsubscribe` to all paths of `leaveGroup`. for example:
```java
        if (isNotInGroup()) {
            if (state == MemberState.FENCED) {
                clearAssignment();
                transitionTo(MemberState.UNSUBSCRIBED);
            }
            subscriptions.unsubscribe();
            return CompletableFuture.completedFuture(null);
        }
```

Also, in the `process(UnsubscribeEvent)` we can call `subscriptions.unsubscribe();` directly if `MembershipMgr` is nonexistent. WDYT?

Reply:
Okay. I revert the change in `MembershipManagerImpl.java#leaveGroup` and call `subscriptions.unsubscribe()` directly in `ApplicationEventProcessor#process` if `heartbeatRequestManager` is not present.